Do not plan INNER join with just filter function

losipiuk opened the pull request:
We want to avoid having INNER JoinNode with just non-equality filter function.
In such case HashJoin execution is planned. And if we have CROSS join followed by
FILTER, NestedLoopsJoin is planned, which is much faster.

kokosing commented on presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java:
```diff
@@ -2656,6 +2656,9 @@ public void testNonEqualityJoin()
         assertQuery(
                 "SELECT COUNT(*) FROM (VALUES 1, 2) t1(a) JOIN (VALUES 10, 11) t2(b) ON a > 2",
                 "VALUES (0)");
+        assertQuery(
```
It does not test anything, you need test plan shape.

losipiuk replied:
Yeah - it is not supposed to. It is just for coverage.
I aggree the proper plan-shape test is missing. Do we really want it provided it does not exist for any other PP functionality?

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java:
```diff
@@ -366,7 +366,7 @@ public PlanNode visitJoin(JoinNode node, RewriteContext<Expression> context)
 
                 // Create new projections for the new join clauses
-                ImmutableList.Builder<JoinNode.EquiJoinClause> joinConditionBuilder = ImmutableList.builder();
+                ArrayList<JoinNode.EquiJoinClause> joinCondition = new ArrayList<>();
```
equiJoinClauses?

losipiuk commented:
@kokosing
* Renamed variable. 
* Fixed/extended failing test `testDistinctLimitOverInequalityJoin` - I hope it still tests what was its original purpose (@martint?).
* Added tests to specifically tests change from this PR

kokosing commented on presto-main/src/test/java/com/facebook/presto/sql/planner/TestLogicalPlanner.java:
```diff
@@ -79,13 +79,58 @@ public void testDistinctLimitOverInequalityJoin()
+                anyTree(
+                        node(DistinctLimitNode.class,
+                                anyTree(
+                                        join(INNER,
```
Any, but not filter.

losipiuk replied:
I do not think it is important for this test. 
Actually maybe we should drop filter from the other query?

kokosing commented on presto-main/src/test/java/com/facebook/presto/sql/planner/TestLogicalPlanner.java:
```diff
@@ -79,13 +79,58 @@ public void testDistinctLimitOverInequalityJoin()
+            throws Exception
+    {
+        assertPlan("SELECT 1 FROM orders o JOIN lineitem l ON o.orderkey < l.orderkey",
+                anyTree(
```
IMO it would be enough to just check if filter node exists or not in plan.

losipiuk replied:
Maybe. This one seems also valid approach to me.

kokosing commented on presto-main/src/test/java/com/facebook/presto/sql/planner/TestLogicalPlanner.java:
```diff
@@ -79,13 +79,58 @@ public void testDistinctLimitOverInequalityJoin()
+            throws Exception
+    {
+        assertPlan("SELECT 1 FROM orders o JOIN lineitem l ON o.shippriority = l.linenumber AND o.orderkey < l.orderkey",
+                anyTree(
```
Any, but not filter.

nezihyigitbasi commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java:
```diff
@@ -366,7 +366,7 @@ public PlanNode visitJoin(JoinNode node, RewriteContext<Expression> context)
 
                 // Create new projections for the new join clauses
-                ImmutableList.Builder<JoinNode.EquiJoinClause> joinConditionBuilder = ImmutableList.builder();
+                ArrayList<JoinNode.EquiJoinClause> equiJoinClauses = new ArrayList<>();
```
`List<JoinNode.EquiJoinClause> equiJoinClauses`

losipiuk commented:
Rebased and squashed. @nezihyigitbasi, if you think it is fine to merge it before release feel free to do that (or I can).

nezihyigitbasi commented:
@losipiuk merged, thanks. Please add this to release notes.
